sdk/monitor/azure-monitor-ingestion/src/main/java/com/azure/monitor/ingestion/LogsIngestionAsyncClient.java

```diff
@@ -0,0 +1,268 @@
+                            .flatMap(result -> result.getErrors().stream()).collect(Collectors.toList()));
+        }
+        return new UploadLogsResult(UploadLogsStatus.SUCCESS, null);
+    }
+
+    private List<byte[]> createGzipRequests(List<Object> logs, ObjectSerializer serializer,
+                                            List<List<Object>> logBatches) {
+        try {
+            List<byte[]> requests = new ArrayList<>();
+            long currentBatchSize = 0;
+            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+
+            JsonGenerator generator = JsonFactory.builder().build().createGenerator(byteArrayOutputStream);
+            generator.writeStartArray();
+            List<String> serializedLogs = new ArrayList<>();
+
+            int currentBatchStart = 0;
+            for (int i = 0; i < logs.size(); i++) {
+                byte[] bytes = serializer.serializeToBytes(logs.get(i));
+                int currentLogSize = bytes.length;
+                currentBatchSize += currentLogSize;
+                if (currentBatchSize > MAX_REQUEST_PAYLOAD_SIZE) {
+                    writeLogsAndCloseJsonGenerator(generator, serializedLogs);
+                    requests.add(gzipRequest(byteArrayOutputStream.toByteArray()));
+
+                    byteArrayOutputStream = new ByteArrayOutputStream();
+                    generator = JsonFactory.builder().build().createGenerator(byteArrayOutputStream);
+                    generator.writeStartArray();
+                    currentBatchSize = currentLogSize;
+                    serializedLogs.clear();
+                    logBatches.add(logs.subList(currentBatchStart, i));
+                    currentBatchStart = i;
+                }
+                serializedLogs.add(new String(bytes, StandardCharsets.UTF_8));
+            }
+            if (currentBatchSize > 0) {
+                writeLogsAndCloseJsonGenerator(generator, serializedLogs);
+                requests.add(gzipRequest(byteArrayOutputStream.toByteArray()));
+                logBatches.add(logs.subList(currentBatchStart, logs.size()));
+            }
+            return requests;
+        } catch (IOException exception) {
+            throw LOGGER.logExceptionAsError(new UncheckedIOException(exception));
+        }
+    }
+
+    private void writeLogsAndCloseJsonGenerator(JsonGenerator generator, List<String> serializedLogs) throws IOException {
+        generator.writeRaw(serializedLogs.stream()
+                .collect(Collectors.joining(",")));
+        generator.writeEndArray();
+        generator.close();
+    }
+
+    /**
+     * Gzips the input byte array.
+     * @param bytes The input byte array.
+     * @return gzipped byte array.
+     */
+    private byte[] gzipRequest(byte[] bytes) {
+        // This should be moved to azure-core and should be enabled when the client library requests for gzipping the
```

Comment:
I agree with moving this to `azure-core` and once it is in `azure-core` this can leverage `AccessibleByteArrayOutputStream` and return `ByteBuffer` instead of `byte[]` and we can optimize away copying the `ByteArrayOutputStream` bytes again